cloudinit/net/ephemeral.py

```diff
@@ -404,48 +407,63 @@ def __init__(
+                ephemeral network setup is skipped.
+        """
         self.interface = interface
         self.ipv4 = ipv4
         self.ipv6 = ipv6
         self.stack = contextlib.ExitStack()
         self.state_msg: str = ""
         self.distro = distro
+        self.connectivity_urls_data = connectivity_urls_data
 
     def __enter__(self):
-        if not (self.ipv4 or self.ipv6):
+        if not self.ipv4 and not self.ipv6:
             # no ephemeral network requested, but this object still needs to
             # function as a context manager
             return self
         exceptions = []
         ephemeral_obtained = False
-        if self.ipv4:
-            try:
-                self.stack.enter_context(
-                    EphemeralDHCPv4(
-                        self.distro,
-                        self.interface,
-                    )
-                )
-                ephemeral_obtained = True
-            except (ProcessExecutionError, NoDHCPLeaseError) as e:
-                LOG.info("Failed to bring up %s for ipv4.", self)
-                exceptions.append(e)
 
-        if self.ipv6:
-            try:
-                self.stack.enter_context(
-                    EphemeralIPv6Network(
-                        self.distro,
-                        self.interface,
-                    )
+        # short-circuit if we already have connectivity to IMDS
+        if _check_connectivity_to_imds(self.connectivity_urls_data):
+            LOG.debug("We already have connectivity to IMDS, skipping DHCP.")
+            return self
+        else:
+            LOG.debug("No connectivity to IMDS, attempting DHCP setup.")
+
+            # first try to bring up ephemeral network for ipv4 (if enabled)
+            # then try to bring up ephemeral network for ipv6 (if enabled)
+            if self.ipv4:
+                ipv4_ephemeral_obtained, ipv4_exception = (
+                    self._perform_ephemeral_network_setup(ip_version="ipv4")
                 )
-                ephemeral_obtained = True
-                if exceptions or not self.ipv4:
-                    self.state_msg = "using link-local ipv6"
-            except ProcessExecutionError as e:
-                LOG.info("Failed to bring up %s for ipv6.", self)
-                exceptions.append(e)
+                ephemeral_obtained |= ipv4_ephemeral_obtained
+                if ipv4_exception:
+                    exceptions.append(ipv4_exception)
+            if self.ipv6:
+                ipv6_ephemeral_obtained, ipv6_exception = (
+                    self._perform_ephemeral_network_setup(ip_version="ipv6")
+                )
+                ephemeral_obtained |= ipv6_ephemeral_obtained
+                if ipv6_exception:
+                    exceptions.append(ipv6_exception)
+
+            # need to set this if we only have ipv6 ephemeral network
+            if not self.ipv4 or not ipv4_ephemeral_obtained:
```

Comment:
I think at this point this should be `if ipv6_ephemeral_obtained and not ipv4_ephemeral_obtained`.

Reply:
hmmmm... fair enough. i shall try that out.